### DISCUSSION TITLE
Inconsistent pressure in NVT runs between LAMMPS and ASE - which one to trust?

### DISCUSSION THREAD

**Discussion by casteln** [mir-group/nequip#345] [Q&A] [answered]
Hello,

When running NVT trajectories of the same system and NequIP model (learned with stress) with ASE and LAMMPS, I get different average results for the pressure.
I find a systematic shift of roughly 200 MPa, with the pressure outputted by LAMMPS always higher.

I've created a minimal reproducible example [available here](https://github.com/casteln/MRE/tree/main/001_nequip-pressure), which details the thermostat and versions used, as well as how the pressure is read from the output files.
It also contains [a figure](https://github.com/casteln/MRE/blob/main/001_nequip-pressure/results/Other_systems.png) that shows that this shift was observed across models (i.e. that it's not an issue with the units).

Would you know where this is coming from? And which value should I trust?

**Comment by Linux-cpp-lisp** [comment]
Hi @casteln ,

Thanks very much for the detailed report!!

Which LAMMPS branch?  One of them lacks stress support and will only give the kinetic term. Sorry about this, I've been meaning to clean up that branch situation for a while now and have not yet have the time, I expect to very soon.

**Reply by casteln** [reply]
Thank you for your quick answer!
I've used the [`stress` branch of pair_nequip](https://github.com/mir-group/pair_nequip/tree/stress).

In the pair_nequip folder I used for the installation, `git show` printed:
```
commit 97f3f7b20c02e964e78365c5c0854faec5ee2480 (HEAD -> stress, origin/stress)
Author: Linux-cpp-lisp <1473644+Linux-cpp-lisp@users.noreply.github.com>
Date:   Sun Dec 11 00:54:46 2022 -0500
```

This install gave me different results than the `main` branch (`main` led to the same pressure value for the same system, volume and temperature regardless of the MLP),  so I thought it was the branch with stress supported.

**Reply by Linux-cpp-lisp** [reply]
Sorry just to make sure I understand--- `stress` branch did or did not resolve the issue?

**Reply by casteln** [reply]
No it did not resolve the issue, the results I shared were already obtained from the stress branch.

**Reply by casteln** [reply]
Any idea on why the pressure output may still be different in the `stress` branch of pair_nequip compared to ASE?

**Reply by youngjae142** [reply]
Does the stress function of the ASE NequIP calculator also include the kinetic term?
LAMMPS calculates the stress as the sum of kinetic and viral terms [[check this out](https://github.com/mir-group/pair_nequip/discussions/14#discussioncomment-2571620)].
The LAMMPS kinetic term is naturally included by LAMMPS itself, while I don't know in the ASE case.

**Reply by casteln** [reply]
Thank you @ssrokyz for your insight!
From [this answer](https://github.com/mir-group/nequip/discussions/323#discussioncomment-5459398) by @Linux-cpp-lisp, it would appear that the stress computed by ASE is only the gradient of the potential energy, and thus do not include the kinetic term.

My understanding is therefore that the difference may come from this kinetic term.
Do you know if it's easy to compute this kinetic term with ASE?

**Reply by Linux-cpp-lisp** [reply, marked as answer]
Hi all, sorry for my late response here and thanks @ssrokyz for chiming in!!

Yes, I would suspect that this is the case. You can see how in the `pair_nequip` unit tests we ask LAMMPS for the stress tensor (not the virial, but still without the temperature / kinetic term) which is what we compare to the output from ASE:
https://github.com/mir-group/pair_nequip/blob/stress/tests/test_python_repro.py#L148-L152
and
https://github.com/mir-group/pair_nequip/blob/stress/tests/test_python_repro.py#L317-L344

**Reply by casteln** [reply]
Thank you for answer!
[I've considered LAMMPS stress output after disabling the kinetic term](https://github.com/casteln/MRE/tree/main/001_nequip-pressure#new-run-for-kinetic-term), and found that it gave the same average pressure value than ASE.

**Reply by Linux-cpp-lisp** [reply]
Great!
